src/test_lib_json/main.cpp

```diff
@@ -3723,8 +3723,8 @@ JSONTEST_FIXTURE_LOCAL(IteratorTest, reverseIterator) {
   std::vector<std::string> values;
   using Iter = decltype(json.begin());
   auto re = std::reverse_iterator<Iter>(json.begin());
-  for (auto it = std::reverse_iterator<Iter>(json.end()); it != re; ++it) {
-    values.push_back(it->asString());
+  for (std::reverse_iterator<Iter> it = std::reverse_iterator<Iter>(json.end()); it != re; ++it) {
```

Comment:
`auto` was sufficient for the type of `it`. There was no need to change this line or the following line.
`it->asString()` should have just worked.

Reply:
```
error: base operand of ‘->’ has non-pointer type ‘std::reverse_iterator<Json::ValueIterator>’
 3727 |     values.push_back(it->asString());
      |                        ^~
```
This is the compile error I get.
I will redo to auto, that was not a change that was needed

Reply:
I can't figure out why -> shouldn't work here.
The changes to reverse_iterator in C++20 shouldn't affect "->"
([cppreference](https://en.cppreference.com/w/cpp/iterator/reverse_iterator))

Reply:
Thanks for sticking with this PR.
I really want to know what's up with your `std::reverse_iterator<Iter>::operator->` implementation.
I'm worried that our `Json::ValueIterator` has some feature or flaw that is causing the problem. It has a fancy inheritance that might be a factor here?

Can you make other `std::reverse_iterator` types and use `->` on them?

Reply:
I had the same idea. And did the following to the test case.
It compiles...
```
struct Json1 {
    int value;
    std::string asString(){
        return "";
    }
};
JSONTEST_FIXTURE_LOCAL(IteratorTest, reverseIterator) {
  Json::Value json;
  json["k1"] = "a";
  json["k2"] = "b";
  std::vector<std::string> values;
  using Iter = decltype(json.begin());
  auto re = std::reverse_iterator<Iter>(json.begin());
  for (auto it = std::reverse_iterator<Iter>(json.end()); it != re; ++it) {
    values.push_back((*it).asString());
  }
  std::vector<Json1> json1;
  std::vector<std::string> values1;
  auto re1 = std::reverse_iterator<decltype(json1.begin())>(json1.begin());

  for (auto it = std::reverse_iterator<decltype(json1.begin())>(json1.end()); it != re1; ++it) {
      values1.push_back(it->asString());
  }
  JSONTEST_ASSERT((values == std::vector<std::string>{"b", "a"}));
}
```

So maybe you are right that Json::ValueIterator is the problem.
One thing they introduced in C++20 is iterator_concept.
Maybe that is where we need to look [iterator_traits](https://en.cppreference.com/w/cpp/iterator/iterator_traits)

Reply:
Can't std::transform be used? It's just a simple push back.

Reply:
> Can't std::transform be used? It's just a simple push back.

Well we could write it in a more elegant way, but remember this isn't an application. It's `JSONTEST_FIXTURE_LOCAL(IteratorTest, reverseIterator)` so we're trying to test our iterator's interaction with `std::reverse_iterator` methodically. We want to really see and review the increment and dereference operations.

Reply:
I figured this out.  The `ValueIterator` `operator->` and `operator*` need to be `const`.

(The constness of the referent has nothing to do with the const qualification on the iterator)